Fix bug with incorrect expiration dates

samjohn opened the pull request:
It looks like Apple changed the way they send receipts such that they no longer appear in order of expiration date. 

This fixes the issue by removing assumptions about the *last* receipt from Apple being the active one. Instead we look for the receipt with the most recent expiration date. 

I also added Byebug and Timecop for testing purposes. Timecop I needed to make the tests pass, byebug I could remove if folks don't like it.

sidonath commented:
I'm not sure this fixes the issue: in [`VerificationResponse#is_subscription_active?`](https://github.com/gabrielgarza/monza/blob/da594ca5df0c49893c96ce349f20590e7d16863f/lib/monza/verification_response.rb#L39-L49) we check `@latest_receipt_info.last` which doesn't come from `@in_app`. I think we should sort `@latest_receipt_info` in [`VerificationResponse#is_subscription_active?`](https://github.com/gabrielgarza/monza/blob/da594ca5df0c49893c96ce349f20590e7d16863f/lib/monza/verification_response.rb#L36)

sidonath commented:
Looking at our receipts that failed verification today, I see only one entry in the `"in_app"` array (matching the earliest subscription cycle), while I see multiple entries in the `"latest_receipt_info"`

samjohn commented:
thanks! This was just a guess--I was about to download some receipts and see what was going on. I'll check it out.

m-o-e commented:
Fwiw, we have applied this Hotfix for now:

```
def latest_expiry_date
  # @latest_receipt_info.last.expires_date_ms if @latest_receipt_info.last
  @latest_receipt_info.collect(&:expires_date_ms).max if @latest_receipt_info.last
end
```

That works for us because `#latest_expiry_date`
is the only expiry-related method we use (all other
methods obv stay broken!).

As for the proper fix, this PR looks good to me
(just ensure the array is sorted according to
what Monza always expected). :+1:

But I could very well be missing something,
so please take that with a grain of salt. 😓

samjohn commented:
@sidonath  was right, I think in_app was the wrong place to look. I updated  `is_subscription_active?`, but will also update all the other places where this uses `latest_receipt_info.last`.

gabrielgarza commented:
Thanks to all, will review this asap and post my comments here soon

gabrielgarza commented on lib/monza/verification_response.rb:
```diff
@@ -37,30 +37,16 @@ def initialize(attributes)
-      else
-        false
-      end
+      is_subscription_active?
```
Hey @sidonath correct, this would change behavior. This is an edge case but happened to me last year, I should've documented it more, but I believe it was due two two separate product subscriptions and one was cancelled but the subscription for the other product was still active

gabrielgarza replied:
@samjohn @sidonath actually I think this is fine as this method will still return true if any, not just the latest, is active

gabrielgarza commented on spec/verification_response_spec.rb:
```diff
@@ -101,12 +113,8 @@ def replace_expires_date(hash, time)
 
     context 'for a receipt with expiration date in the future' do
       let(:verify) do
-        response["receipt"]["in_app"].each do |in_app|
```
@samjohn why should we delete this part for `in_app`?

samjohn replied:
oh, I did this because in this test *every* receipt has an expiration date in the future so the order doesn't matter. this way you just make the first one in the future. I think your solution was more clear though.

gabrielgarza commented on lib/monza/verification_response.rb:
```diff
@@ -70,7 +56,7 @@ def latest_active_transaction_receipt
     end
 
     def latest_expiry_date
-      @latest_receipt_info.last.expires_date_ms if @latest_receipt_info.last
```
@sidonath  could we change this so it supports Ruby 2.2?

gabrielgarza commented:
@samjohn I agree with these changes, I believe you could have also just used `is_any_subscription_active?` as it checks for an active subscription in the array, not just the last item. Regardless, I think these changes improve the gem and make it less fragile.

I left my comments / questions above @samjohn , thank you!

samjohn commented:
Sorry! didn't have time to look at your comments this week--but looks like you have a fix. Thank you @gabrielgarza !

samjohn commented:
@gabrielgarza  I updated this PR to look more like your changes stylistically and added code to ensure that the gem can handle receipts without an expiration date.

gabrielgarza commented:
@samjohn thank you for updating the branch with the fixes and looking at the other branch I uploaded. Will merge this now and push

gabrielgarza commented:
@samjohn @sidonath @m-o-e @benrugg just pushed 1.3.7 out to Rubygems, thank you!

benrugg commented:
Wonderful, thank you!

m-o-e commented:
🙇
